# DS 3000 - Assignment [2]

**Student Name**: [Brian Wong]

**Date**: [1/14/2025]


### Submission Instructions
Submit this `ipynb` file to canvas.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run `Kernel > Restart & Run All` just before uploading the `ipynb` file to Canvas.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or DM us in MS Teams if you have a specific question about your work or if you would like another pair of eyes or talk through your code.

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment. You do not need to cite the official python documentation.

**Documentation / style counts for credit**  Please refer to the Pep-8 style, to improve the readability and consistency of your Python code. For more information, read the following article [How to Write Beautiful Python Code With PEP 8](https://realpython.com/python-pep8/) or ask your TA's for tips.

**NOTE:<span style='color:red'> Write python expressions to answer ALL questions below and ensure that you use the `print()` function to display the output.</span>** Each question should be answered in a new code cell. For example, your solution for question 1.1 should be in a different code cell from your solution for question 1.2.

## Question 1: Analyzing numeric Data (50 pts) 

TODO: The data that you are working with pertains to the risk factors that are associated with low infant birth weight**. The data contains the following fields:

low:
indicator of birth weight less than 2.5 kg.

age:
mother's age in years.

lwt:
mother's weight in pounds at last menstrual period.

race:
mother's race (1 = white, 2 = black, 3 = other).

smoke:
smoking status during pregnancy  (0 = false, 1 = true).

ptl:
number of previous premature labors.

ht:
history of hypertension (0 = false, 1 = true).

ui:
presence of uterine irritability.

ftv:
number of physician visits during the first trimester.

bwt:
birth weight in grams.

Write python code to answer the questions below and ensure that you <b>round all numeric calculations (i.e. floats) to 2 decimal places (unless indicated otherwise).</b>
Tip: you will need to filter for conditions that are true before performing certain calculations.


In [3]:
import numpy as np

## Question 1.1: 
TODO: (5 pts) Load the attached data into a numpy array and display the dimensions of the data : birthwt.csvDownload birthwt.csv
Note: do not count the column headers, only the data.


In [5]:
data = np.loadtxt('birthwt.csv',
                 delimiter = ',',
                 skiprows = 1)
print(data.shape, " 189 rows and 10 columns, excluding column headers")

(189, 10)  189 rows and 10 columns, excluding column headers


## Question 1.2: 
TODO: (3 pts) What is the age range of mothers in this dataset.
Note: the range should include: the minimum, average and maximum age. Ensure that the ages are represented as whole numbers.

In [7]:
min_age = np.min(data[:,1])
print('The minimum age of mothers in this dataset is: ', int(min_age))
max_age = np.max(data[:,1])
print('The maximum age of mothers in this dataset is: ', int(max_age))
avg_age = round(np.mean(data[:,1]))
print('The average age of mothers in this dataset is: ', int(avg_age))
range = max_age - min_age
print('The age range of mothers in this dataset is: ', int(range))

The minimum age of mothers in this dataset is:  14
The maximum age of mothers in this dataset is:  45
The average age of mothers in this dataset is:  23
The age range of mothers in this dataset is:  31


## Question 1.3: 
TODO: (2 pts) How many mothers smoked during pregnancy?

In [9]:
smoke = data[:,4]
num_smoke = smoke[smoke == 1]
length = len(num_smoke)
print(length, " mothers smoked during pregnancy.")

74  mothers smoked during pregnancy.


## Question 1.4: 
TODO: (2 pts) How many mothers had a history of hypertension?

In [11]:
ht = data[:,6]
num_ht = ht[ht == 1]
length = len(num_ht)
print(length, " mothers had a history of hypertension.")

12  mothers had a history of hypertension.


## Question 1.5: 
TODO: (3 pts) How many mothers have 1 or more premature births?

In [13]:
ptl = data[:,5]
num_ptl = ptl[ptl >= 1]
length = len(num_ptl)
print(length, " mothers have 1 or more premature births.")

30  mothers have 1 or more premature births.


## Question 1.6: 
TODO: (5 pts) Create a function called convert_birthweight() that:
- takes grams as input,
- converts grams to the equivalent pounds and ounces
example format: X grams = Y lbs Z oz; note: X, Y and Z are placeholders for numeric values.
- return the converted value.
Use the convert_birthweight() function to convert the birth weight to pounds and ounces in the questions 7-11 below.

In [15]:
def convert_birthweight(grams):
    grams_per_pound = 453.59237
    ounces_per_pound = 16
    pounds = grams / grams_per_pound
    whole_pounds = int(pounds)
    remaining_ounces = (pounds - whole_pounds) * ounces_per_pound
    whole_ounces = round(remaining_ounces, 2)
    return f"{whole_pounds} pounds and {whole_ounces} ounces"

## Question 1.7: 
TODO: (5 pts) What is the average birth weight of babies whose mothers have a history of premature birth? 

In [17]:
premature_birth = data[data[:,5] > 0]
birth_weights = premature_birth[:,9]
avg_birth_weight = np.mean(birth_weights)
result_gram = np.round(avg_birth_weight, 2)
result_converted = convert_birthweight(avg_birth_weight)
print(result_gram, 'grams, or', result_converted, 'is the average birth weight of babies whose mothers have a history of premature birth')

2579.4 grams, or 5 pounds and 10.99 ounces is the average birth weight of babies whose mothers have a history of premature birth


## Question 1.8: 
TODO: (5 pts) What is the minimum, average and maximum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension?

In [19]:
no_smoke_and_no_ht = data[(data[:,4] == 0) & (data[:,6] == 0)]
birth_weights = no_smoke_and_no_ht[:,9]

min_birth_weight_gram = np.min(birth_weights)
avg_birth_weight_gram = np.mean(birth_weights)
max_birth_weight_gram = np.max(birth_weights)

avg_gram_rounded = np.round(avg_birth_weight_gram, 2)

min_birth_weight_converted = convert_birthweight(min_birth_weight_gram)
avg_birth_weight_converted = convert_birthweight(avg_birth_weight_gram)
max_birth_weight_converted = convert_birthweight(max_birth_weight_gram)

print(min_birth_weight_gram, 'grams, or', min_birth_weight_converted, 'is the minimum birth weight of babies whose mother ' 
      +'did not smoke during pregnancy and do not have a history of hypertension.')
print(avg_gram_rounded, 'grams, or', avg_birth_weight_converted, 'is the average birth weight of babies whose mother ' 
      +'did not smoke during pregnancy and do not have a history of hypertension.')
print(max_birth_weight_gram, 'grams, or', max_birth_weight_converted, 'is the maximum birth weight of babies whose mother ' 
      +'did not smoke during pregnancy and do not have a history of hypertension.')

1021.0 grams, or 2 pounds and 4.01 ounces is the minimum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension.
3090.44 grams, or 6 pounds and 13.01 ounces is the average birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension.
4990.0 grams, or 11 pounds and 0.02 ounces is the maximum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension.


## Question 1.9: 
TODO: (5 pts) What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy?

In [21]:
smoked = data[data[:,4] == 1]
birth_weights = smoked[:,9]

min_birth_gram = np.min(birth_weights)
avg_birth_gram = np.mean(birth_weights)
max_birth_gram = np.max(birth_weights)

avg_gram_rounded = np.round(avg_birth_gram, 2)

min_birth_gram_converted = convert_birthweight(min_birth_gram)
avg_birth_gram_converted = convert_birthweight(avg_birth_gram)
max_birth_gram_converted = convert_birthweight(max_birth_gram)

print(min_birth_gram, 'grams, or', min_birth_gram_converted, 'is the minimum birth weight of babies whose mothers smoked during pregnancy.')
print(avg_gram_rounded, 'grams, or', avg_birth_gram_converted, 'is the average birth weight of babies whose mothers smoked during pregnancy.')
print(max_birth_gram, 'grams, or', max_birth_gram_converted, 'is the maximum birth weight of babies whose mothers smoked during pregnancy.')

709.0 grams, or 1 pounds and 9.01 ounces is the minimum birth weight of babies whose mothers smoked during pregnancy.
2771.92 grams, or 6 pounds and 1.78 ounces is the average birth weight of babies whose mothers smoked during pregnancy.
4238.0 grams, or 9 pounds and 5.49 ounces is the maximum birth weight of babies whose mothers smoked during pregnancy.


## Question 1.10: 
TODO: (5 pts) What is the minimum, average and maximum birth weight of babies whose mothers had a history of hypertension? 

In [23]:
ht = data[data[:,6] == 1]
birth_weights = ht[:,9]

min_birth_gram = np.min(birth_weights)
avg_birth_gram = np.mean(birth_weights)
max_birth_gram = np.max(birth_weights)

avg_gram_rounded = np.round(avg_birth_gram, 2)

min_birth_gram_converted = convert_birthweight(min_birth_gram)
avg_birth_gram_converted = convert_birthweight(avg_birth_gram)
max_birth_gram_converted = convert_birthweight(max_birth_gram)

print(min_birth_gram, 'grams, or', min_birth_gram_converted, 'is the minimum birth weight of babies whose mothers had a history of hypertension.')
print(avg_gram_rounded, 'grams, or', avg_birth_gram_converted, 'is the average birth weight of babies whose mothers had a history of hypertension.')
print(max_birth_gram, 'grams, or', max_birth_gram_converted, 'is the maximum birth weight of babies whose mothers had a history of hypertension.')

1135.0 grams, or 2 pounds and 8.04 ounces is the minimum birth weight of babies whose mothers had a history of hypertension.
2536.83 grams, or 5 pounds and 9.48 ounces is the average birth weight of babies whose mothers had a history of hypertension.
3790.0 grams, or 8 pounds and 5.69 ounces is the maximum birth weight of babies whose mothers had a history of hypertension.


## Question 1.11: 
TODO: (5 pts) What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy <b>and</b> had a history of hypertension? 

In [25]:
smoke_and_ht = data[(data[:,4] == 1) & (data[:,6] == 1)]
birth_weights = smoke_and_ht[:,9]

min_birth_weight_gram = np.min(birth_weights)
avg_birth_weight_gram = np.mean(birth_weights)
max_birth_weight_gram = np.max(birth_weights)

avg_gram_rounded = np.round(avg_birth_weight_gram, 2)

min_birth_weight_converted = convert_birthweight(min_birth_weight_gram)
avg_birth_weight_converted = convert_birthweight(avg_birth_weight_gram)
max_birth_weight_converted = convert_birthweight(max_birth_weight_gram)

print(min_birth_weight_gram, 'grams, or', min_birth_weight_converted, 'is the minimum birth weight of babies whose mother ' 
      +'smoked during pregnancy and had a history of hypertension.')
print(avg_gram_rounded, 'grams, or', avg_birth_weight_converted, 'is the average birth weight of babies whose mother ' 
      +'smoked during pregnancy and had a history of hypertension.')
print(max_birth_weight_gram, 'grams, or', max_birth_weight_converted, 'is the maximum birth weight of babies whose mother ' 
      +'smoked during pregnancy and had a history of hypertension.')

1135.0 grams, or 2 pounds and 8.04 ounces is the minimum birth weight of babies whose mother smoked during pregnancy and had a history of hypertension.
2561.0 grams, or 5 pounds and 10.34 ounces is the average birth weight of babies whose mother smoked during pregnancy and had a history of hypertension.
3756.0 grams, or 8 pounds and 4.49 ounces is the maximum birth weight of babies whose mother smoked during pregnancy and had a history of hypertension.


## Question 1.12: 
TODO: (5 pts) Compare and contrast the results from 7-11 above and summarize your findings.
Do you see any patterns in the birth weight of babies whose mother did not have any history of hypertension and/or did not smoke during pregnancy versus those who did? Ensure that you reference specific birth weights to support your explanation.

I noticed that the birth weight of babies whose mothers did not smoke during pregnancy and did not have a history of
hypertension had the highest average birth weight at 6 pounds and 13.01 ounces compared to the birth weight of babies whose mothers smoked during
pregancy, had a history of hypertension, and smoked and had a history of hypertension at 6 pounds and 1.78 ounces, 5 pounds and 9.48 ounces, and
5 pounds and 10.34 ounces respectively. 

Based on these average weights, I also noticed the average birth weight of babies whose mothers have a history of premature birth, had a history of hypertension, and smoked during pregnancy and had a history of hypertension all had similar weights at 5 pounds and 10.99 ounces, 5 pounds and 9.48 ounces, and 5 pounds and 10.34 ounces respectively.

Another pattern that I noticed is that the minimum, average, and maximum birth weight of babies whose mothers had a history of hypertension--at 2 pounds and 8.04 ounces, 5 pounds and 9.48 ounces, and 8 pounds and 5.69 ounces respectively--compared to those of mothers who smoked during pregnancy and had a history of hypertension--at 2 pounds and 8.04 ounces, 5 pounds and 10.34 ounces, and 8 pounds and 4.49 ounces respectively--had very similar weights.

Another pattern I noticed is that the minimum birth weight of babies was the lowest for mothers who smoked during pregnancy at 1 pound and 9.01 ounces compared to the minimum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension, whose mothers had a history of hypertension, and whose mother smoked during pregnancy and had a history of hypertension at 2 pounds and 4.01 ounces, 2 pounds and 8.04 ounces, and 2 pounds and 8.04 ounces respectively.

In terms of the maximum birth weight of babies, the highest was for those whose mother did not smoke during pregnancy and do not have a history of hypertension at 11 pounds and 0.02 ounces compared to the maximum birth weight of babies whose mothers smoked during pregnancy, whose mothers had a history of hypertension, and whose mother smoked during pregnancy and had a history of hypertension at 9 pounds and 5.49 ounces, 8 pounds and 5.69 ounces, and 8 pounds and 4.49 ounces respectively.

Overall, I noticed that mothers who neither smoked nor had hypertension gave birth to babies with higher averages and maximum birth weights, suggesting that these factors may improve birth weights.

## Helpful resources 
Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment inside the code cell, or you can list them below. 

You do not need to cite the official python documentation.
